syft/frameworks/torch/hook/hook_args.py

```diff
@@ -55,6 +55,7 @@
     "chunk",
     "torch.functional.split",
     "split",
+    "backward",
```

Comment:
why is this an ambiguous function but not an ambiguous method?  I also feel we don't need it to be ambiguous since this will commonly be done in a loop with the same or very similar args (however I'm not entirely sure how similar signatures need to be for this to save/waste time).

Reply:
> why is this an ambiguous function but not an ambiguous method? 
Yes excellent point!
Thats because of the response type, not the input type! For response wrapping we use "ambiguous function" (see hook_response and register_response)

>  this will commonly be done in a loop with the same or very similar args
Indeed, but in a same session I might want to do FL AND encrypted training, which fails, as pointed in #2503

Reply:
>Indeed, but in a same session I might want to do FL AND encrypted training, which fails, as pointed in #2503

Hm, I'm not sure why that wouldn't work -- I was under the impression that ambiguous functions/methods are only different because of their signature type, and moreover that they were entirely optional.  If the signature changes, I feel like that should be noticed lazily, i.e. if I call an ambiguous function 1000 times in a row with the same signature and then call it once more after with a new signature, we shouldn't have to rebuild the signature the first 1000 times.  Maybe I'm missing something though, I'm still new to this part of the code!

Reply:
>  Further, I thought backward always returns None -- I must be missing something?

Theoretically that's right. However with distant commands we do pre-allocation of responses: meaning that we create a pointer to the response before the command is sent and executed. If the response contained more than 1 tensor, we have a fallback mechanism and we interactively create several pointers. But if the response is empty, no such mechanism exist to limit communication: you end up with a stale pointer, and usually people don't try to play with it (like no one does ptr = x.backward() ). This stale pointer goes in register_response.

> I thought these ambiguous functions/methods were entirely optional.

They are not. hook_args has a failure mode when the signature contains more tensors than expected, **but not when are are less than expected**. This is for efficiency purpose, as making length comparison all over the calls would certainly kill the benefit from having hook_args. Therefore, functions with changing signatures have to be registered as *ambiguous*

Reply:
Got it, that makes sense.  Thank you for clarifying!

Reply:
That was very helpful thanks @LaRiffle